src/libsyntax/parse/parser.rs

```diff
@@ -5232,23 +5234,32 @@ impl<'a> Parser<'a> {
             }
             let mut err = self.diagnostic().struct_span_err(id_sp,
                 "cannot declare a new module at this location");
-            let this_module = match self.directory.path.file_name() {
-                Some(file_name) => file_name.to_str().unwrap().to_owned(),
-                None => self.root_module_name.as_ref().unwrap().clone(),
-            };
-            err.span_note(id_sp,
-                          &format!("maybe move this module `{0}` to its own directory \
-                                     via `{0}/mod.rs`",
-                                    this_module));
+            if id_sp != syntax_pos::DUMMY_SP {
+                let mut src_path = PathBuf::from(self.sess.codemap().span_to_filename(id_sp));
+                let src_path_copy = src_path.clone();
+                let stem = src_path_copy.file_stem().unwrap_or(OsStr::new(""));
```

Comment:
If there's no `file_stem`, I think we should use the full file name or just not emit the note. I believe using `""` here would produce a very confusing suggestion.